java/src/main/scala/prometheus4cats/javasimpleclient/JavaMetricRegistry.scala

```diff
@@ -63,8 +66,51 @@ class JavaMetricRegistry[F[_]: Async: Logger] private (
     case _ => name.show
   }
 
-  private def timeoutCallback[A](fa: F[A], onError: A, stringName: String): A =
-    Utils.timeoutCallback(dispatcher, callbackTimeout, fa, onError, s"Affected metric: '$stringName'")
+  private def trackErrors[A](state: Ref[F, Set[String]], stringName: String, onContains: F[A], onContainsNot: F[A]) =
+    state.modify { current =>
+      if (current.contains(stringName)) (current, onContains) else (current + stringName, onContainsNot)
+    }.flatten
+
+  private def incErrorCounter(stringName: String, reason: String) =
+    Sync[F].delay(callbackErrorCounter.labels(stringName, reason).inc())
+
+  private def timeoutCallback[A](fa: F[A], empty: A, stringName: String): A = {
+    lazy val incTimeout = incErrorCounter(stringName, "timeout").as(empty)
+    lazy val incError = incErrorCounter(stringName, "error").as(empty)
+
+    Utils.timeoutCallback(
+      dispatcher,
+      callbackTimeout,
+      fa,
+      th =>
+        trackErrors(
+          callbackTimeoutState,
+          stringName,
+          incTimeout,
+          Logger[F].warn(th)(
+            s"Timed out running callback for metric '$stringName' after $callbackTimeout.\n" +
+              "This may be due to a callback having been registered that performs some long running calculation which blocks\n" +
+              "Please review your code or raise an issue or pull request with the library from which this callback was registered.\n" +
+              s"This warning will only be shown once for each metric after process start. The counter '${JavaMetricRegistry.callbackErrorCounterName}'" +
+              "tracks the number of times this occurs per metric name."
+          ) >> incTimeout
```

Comment:
technically this should probably be `.guarantee`, although it's highly unlikely to make a difference